### PR TITLE
Reduce Uri ctor overhead

### PR DESCRIPTION
**Pull request by MihaZupan**
Reduces the initial overhead for every Uri ctor.

```c#
public class UriInitOverhead
{
    [Benchmark]
    public Uri Ascii() => new Uri("http://foo/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa");

    [Benchmark]
    public Uri PercentEncoded() => new Uri("http://foo/%01%02%03%04%0F%2F%3F%7F%01%02%03%04%0F%2F%3F%7F");
}
```

|         Method |              Toolchain |     Mean |   Error |  StdDev | Ratio |
|--------------- |----------------------- |---------:|--------:|--------:|------:|
|          Ascii | \perf-base\CoreRun.exe | 226.3 ns | 0.42 ns | 0.33 ns |  1.42 |
|          Ascii |  \perf-new\CoreRun.exe | 159.6 ns | 1.63 ns | 1.53 ns |  1.00 |
|                |                        |          |         |         |       |
| PercentEncoded | \perf-base\CoreRun.exe | 499.3 ns | 1.13 ns | 1.00 ns |  2.62 |
| PercentEncoded |  \perf-new\CoreRun.exe | 190.3 ns | 0.52 ns | 0.44 ns |  1.00 |

Will also improve perf for any processing that calls into `UriHelper.EscapedAscii` - anywhere we unescape percent-encoded values.

### CODE REVIEW

**Comment by gfoidl** [comment]
Should this benchmarks be incorporated to [Perf_Uri](https://github.com/dotnet/performance/blob/5db891b26e87bd29617b612cd9ec5c5e22f15e96/src/benchmarks/micro/libraries/System.Runtime/Perf.Uri.cs#L11)?

**Comment by MihaZupan** [comment]
Yes, the list of Uri-related benchmarks should be extended in the perf repo

File: src/libraries/System.Private.Uri/src/System/UriHelper.cs [hunk 1]
```diff
@@ -668,35 +668,36 @@ internal static void EscapeAsciiChar(byte b, ref ValueStringBuilder to)
             HexConverter.ToCharsBuffer(b, to.AppendSpan(2), 0, HexConverter.Casing.Upper);
         }
 
-        internal static char EscapedAscii(char digit, char next)
+        internal static char EscapedAscii(uint digit, uint next)
```
Comment (danmosemsft) [inline src/libraries/System.Private.Uri/src/System/UriHelper.cs:671]: Might be worth a comment indicating what this method does, and maybe a better name. Eg. 
```c#
// Converts 2 hex chars to a byte (returned in a char) 
// e.g, '0' 'a' becomes (char)0x0A
// If either char is not hex, returns (char)0xFFFF
internal static char UnescapeHexChars(uint digit, uint next)
```
Unless I'm misunderstanding.

Comment (MihaZupan) [reply src/libraries/System.Private.Uri/src/System/UriHelper.cs]: I've renamed it to `DecodeHexChars` and added a comment

**Comment by MihaZupan** [comment]
The change in the hex decoding also improves perf for `Uri.UnescapeDataString`

```c#
[Benchmark]
public string UnescapeDataString_ReservedAscii() => Uri.UnescapeDataString("%3F%20%3E%21%3F%20%3E%21%3F%20%3E%21%3F%20%3E%21%3F%20%3E%21");
```

|              Toolchain |     Mean |   Error |  StdDev |   Median | Ratio | RatioSD |
|----------------------- |---------:|--------:|--------:|---------:|------:|--------:|
| \perf-base\CoreRun.exe | 230.9 ns | 2.51 ns | 3.60 ns | 229.3 ns |  1.20 |    0.02 |
|  \perf-new\CoreRun.exe | 192.5 ns | 3.62 ns | 5.30 ns | 196.7 ns |  1.00 |    0.00 |

File: src/libraries/System.Private.Uri/src/System/UriExt.cs [hunk 2]
```diff
@@ -213,22 +212,24 @@ private void InitializeUri(ParsingError err, UriKind uriKind, out UriFormatExcep
...
+        // Also checks for sequences that are 3986 Unreserved characters as these should be un-escaped
+        private static bool CheckForUnicodeOrEscapedUnreserved(string data)
         {
             for (int i = 0; i < data.Length; i++)
             {
                 char c = data[i];
                 if (c == '%')
                 {
-                    if (i + 2 < data.Length)
+                    if ((uint)(i + 2) < (uint)data.Length)
```
Comment (stephentoub) [inline src/libraries/System.Private.Uri/src/System/UriExt.cs:224]: Same question as on the other PR; does this uint casting actually enable the JIT to eliminate the i+1 and i+2 bounds checks?  If it does, great; I just didn't think it did.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.Uri/src/System/UriExt.cs (hunk 2 above)
Comment (MihaZupan) [reply src/libraries/System.Private.Uri/src/System/UriExt.cs:224]: I measured it against a completely unsafe version and the perf is: (`new` being the unsafe variant) 

|         Method |              Toolchain |     Mean |   Error |  StdDev |   Median | Ratio | RatioSD |
|--------------- |----------------------- |---------:|--------:|--------:|---------:|------:|--------:|
|          Ascii | \perf-base\CoreRun.exe | 157.4 ns | 0.26 ns | 1.33 ns | 157.0 ns |  0.95 |    0.01 |
|          Ascii |  \perf-new\CoreRun.exe | 165.7 ns | 0.34 ns | 1.68 ns | 165.5 ns |  1.00 |    0.00 |
|                |                        |          |         |         |          |       |         |
| PercentEncoded | \perf-base\CoreRun.exe | 192.7 ns | 1.94 ns | 9.67 ns | 186.9 ns |  0.98 |    0.05 |
| PercentEncoded |  \perf-new\CoreRun.exe | 196.7 ns | 0.25 ns | 1.22 ns | 196.6 ns |  1.00 |    0.00 |

That said, I'll merge it as-is

**Pull request merged**